db/src/wal.rs

```diff
@@ -0,0 +1,302 @@
+use anyhow::Result;
+use parity_scale_codec::{Decode, Encode, IoReader};
+use std::borrow::Cow;
+use std::fmt::Debug;
+use std::fs::{self, File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::time::{SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error("this directory doesn't contain a wal")]
+pub struct InvalidWal;
+
+/// A wal entry.
+///
+/// Entries are 128 bytes. They have the format
+/// u8 -> payload -> zero padding
+enum Entry<'a, T: Clone + 'static> {
+    /// An operation of a transaction.
+    Op(Cow<'a, T>),
+    /// Begins executing the transaction.
+    ///
+    /// Contains the next id counter after successfully executing the
+    /// transaction.
+    Begin(u64),
+    /// Marks a transaction as executed or aborted.
+    ///
+    /// Contains the last next id counter if the transaction was aborted
+    /// or the new next id counter if it was executed.
+    End(u64),
```

Comment:
So we only know what this inner `u64` refers to if we know the execution state of the transaction. Probably fine, but it may be more clear to separate this into two variants: (1) `Aborted(u64)` which implies the inner `u64` is the last next id counter and (2) `Executed(u64)` which implies the inner `u64` is the new next id counter.

Reply:
It doesn't really matter what the execution state is...